### PR TITLE
Tap Detection Update

### PR DESCRIPTION
**Pull request by finger563**
PR for issue #320 

#### Implement:
- [x] Ensure the features requested from #320 have been implemented

#### Testing:
- [x] Test hard taps
- [x] Test fast taps
- [x] Test tap detection latency (#371 )
- [x] Ensure code refactored for readability / OOP / streamlining

@p-ranav  you've not had extensive training / experience with the SD yet, so you'll be a good equivalent of a new user with little training for determining how effective the current tap detection is for new users

#### Sign-Off:
Ben Needs to Sign off on:
- [x] Tap detection update behavior (hard, fast, location on wrist, looseness of the band, etc.)
- [x] Default tap sensitivity
- [x] Tap sensitivity controls
- [x] Motor activation tap sensitivity
- [x] Motor deactivation tap sensitivity

### CODE REVIEW

File: apps/wear/smartdrive/app/models/tap-detector.ts [hunk 1]
```diff
@@ -151,12 +155,23 @@ export class TapDetector {
...
    */
-  public setSensitivity(sensitivity: number) {
+  public setSensitivity(sensitivity: number, motorOn: boolean) {
     // ensure sensitivity is in range [0, 100]
     sensitivity = Math.min(100, Math.max(sensitivity, 0));
     // update jerk threshold
     this.jerkThreshold = this.maxJerkThreshold - (this.maxJerkThreshold - this.minJerkThreshold) * (sensitivity / 100.0);
     this.predictionThreshold = this.maxPredictionThreshold - (this.maxPredictionThreshold - this.minPredictionThreshold) * (sensitivity / 100.0);
+
+    //harder to tap start and easier to tap stop the motor
```
Comment (p-ranav) [inline apps/wear/smartdrive/app/models/tap-detector.ts]: Add space after `//`

File: apps/wear/smartdrive/app/models/tap-detector.ts [hunk 2]
```diff
@@ -12,13 +12,17 @@ export class TapDetector {
...
-  private maxPredictionThreshold = 0.8;
+  private minPredictionThreshold: number = 0.5;
+  private maxPredictionThreshold: number = 1.0;
   private predictionThreshold: number = 0.5; // confidence
+  private predictionThresholdOnOffDiff: number = 0.2;
+  private predictionThresholdDynamic: number = 0.5;
 
-  private jerkThreshold: number = 17.0; // acceleration value
-  private maxJerkThreshold: number = 30.0;
-  private minJerkThreshold: number = 17.0;
```
Comment (p-ranav) [inline apps/wear/smartdrive/app/models/tap-detector.ts]: Need some comments here. What is `jerkThreshold`? What is `jerkThresholdDynamic`? What is `OnOffDiff`?

File: apps/wear/smartdrive/app/models/tap-detector.ts [hunk 3]
```diff
@@ -227,28 +242,37 @@ export class TapDetector {
...
       jerkAboveThreshold && predictionsWereGood && this.tapGap;
+
+    // exceptions
+    const predictionsOvermin = this.predictionHistory.reduce((good, prediction) => {
+      return good && prediction > this.minPredictionThreshold;
+    }, true);
+    const mustTap = maxJerk > this.maxJerkThreshold && predictionsOvermin;
     // record that there has been a tap
-    if (predictTap) {
+    const realTap=predictTap || mustTap;
```
Comment (p-ranav) [inline apps/wear/smartdrive/app/models/tap-detector.ts]: Need space on either side of `=`

File: apps/wear/smartdrive/app/models/tap-detector.ts [hunk 4]
```diff
@@ -151,12 +155,23 @@ export class TapDetector {
...
     this.predictionThreshold = this.maxPredictionThreshold - (this.maxPredictionThreshold - this.minPredictionThreshold) * (sensitivity / 100.0);
+
+    //harder to tap start and easier to tap stop the motor
+    if (!motorOn) {
+      this.jerkThresholdDynamic = this.jerkThreshold;
+      this.predictionThresholdDynamic = this.predictionThreshold;
+    }
+    else {
+      this.jerkThresholdDynamic = this.jerkThreshold - this.jerkThresholdOnOffDiff;
+      this.predictionThresholdDynamic=this.predictionThreshold - this.predictionThresholdOnOffDiff;
```
Comment (p-ranav) [inline apps/wear/smartdrive/app/models/tap-detector.ts]: Need space on either side of `=`

File: apps/wear/smartdrive/app/models/tap-detector.ts [hunk 5]
```diff
@@ -52,8 +56,8 @@ export class TapDetector {
    * Higher-level tap detection - not TFLite related
    */
   private static InputHistorySize = 4;
-  private static PredictionHistorySize = 1;
-  private static InputRawHistorySize = 10;
+  private static PredictionHistorySize = 2;
+  private static InputRawHistorySize = 6;
```
Comment (p-ranav) [inline apps/wear/smartdrive/app/models/tap-detector.ts]: Might be redundant but for consistency, specify the datatype `: number` for these members.

File: apps/wear/smartdrive/app/models/tap-detector.ts [hunk 6]
```diff
@@ -151,12 +155,23 @@ export class TapDetector {
...
+    // harder to tap start and easier to tap stop the motor
+    if (!motorOn) {
+      this.jerkThresholdDynamic = this.jerkThreshold;
+      this.predictionThresholdDynamic = this.predictionThreshold;
+    }
+    else {
+      this.jerkThresholdDynamic = this.jerkThreshold - this.jerkThresholdOnOffDiff;
+      this.predictionThresholdDynamic = this.predictionThreshold - this.predictionThresholdOnOffDiff;
+    }
+
```
Comment (finger563) [inline apps/wear/smartdrive/app/models/tap-detector.ts]: We should test this thoroughly at different sensitivity settings to see if we want a constant offset (as it is) or if it would be better to make it percentage / multiplicative based change.

**Comment by finger563** [comment]
Prior to merging we also need a write-up which details the tap detection criteria / procedure for review by the rest of  the team.

File: apps/wear/smartdrive/app/models/tap-detector.ts (hunk 6 above)
Comment (xing-w) [reply apps/wear/smartdrive/app/models/tap-detector.ts]: All the pre-defined numbers, even data history sizes, need to be adjusted carefully after thoroughly test.

Comment (finger563) [reply apps/wear/smartdrive/app/models/tap-detector.ts]: Can you elaborate more on this? What were the test results which led you to that conclusion? Please work with @p-ranav on testing the tap detection and adjusting / documenting the algorithm and the proposed changes.

Comment (p-ranav) [reply apps/wear/smartdrive/app/models/tap-detector.ts]: @xing-w Let me know how I can help.

File: apps/wear/smartdrive/app/pages/main-page/main-view-model.ts [hunk 7]
```diff
@@ -1204,17 +1202,20 @@ export class MainViewModel extends Observable {
         return element1;
       });
 
-      const averageAccel = {
+      const averageAccel: Acceleration = {
```
Comment (finger563) [inline apps/wear/smartdrive/app/pages/main-page/main-view-model.ts]: We should probably rename this variable since it's no longer the average... it seems to be the signed absolute maximum (e.g. min if < 0, max if > 0)

File: apps/wear/smartdrive/app/pages/main-page/main-view-model.ts [hunk 8]
```diff
@@ -1204,17 +1202,20 @@ export class MainViewModel extends Observable {
...
         x: total.accel.x >= 0 ? max.accel.x : min.accel.x,
         y: total.accel.y >= 0 ? max.accel.y : min.accel.y,
         z: total.accel.z >= 0 ? max.accel.z : min.accel.z,
       };
 
       const averageTimestamp = total.timestamp / this._previousDataLength;
+      if (((android.os.SystemClock.elapsedRealtimeNanos() - averageTimestamp) / 1000000) > 100) {
+        Log.E('time diff:', ((android.os.SystemClock.elapsedRealtimeNanos() - averageTimestamp) / 1000000));
+      }
```
Comment (finger563) [inline apps/wear/smartdrive/app/pages/main-page/main-view-model.ts]: We'll want to comment this whole section out so that we're not doing unnecessary calls to system clock and logging in the final system.

File: apps/wear/smartdrive/app/services/sqlite.service.ts [hunk 9]
```diff
@@ -5,7 +5,7 @@ function _exists(o, k): boolean {
   return o[k] !== undefined && o[k] !== null && o[k] !== NaN;
 }
 
-const enableMultithreading: boolean = false;
+const enableMultithreading: boolean = true;
```
Comment (finger563) [inline apps/wear/smartdrive/app/services/sqlite.service.ts]: I think we want to revert this change since we determined that it did not affect performance? @xing-w  can you do the logging of the timing test again after setting this back to false to confirm? If we don't see a difference in the maximum latency then we should revert this change.

Comment (xing-w) [reply apps/wear/smartdrive/app/services/sqlite.service.ts]: No affect. We could change it back to false.

**Comment by p-ranav** [comment]
# Tap Detector

## How do we detect taps?

We use the device accelerometer to detect taps. Our input data looks like this:

`[ [acceleration.x, acceleration.y, acceleration.z] ]`

Our tap detector machine learning model takes 2 inputs:
1. Current input, as seen above
2. Previous state (History) - An array of floats of size `TapDetector.StateSize` (currently set to 128) of predictions made by the tap detector so far.

So, the shape of this input is [3, 128].

The model outputs a prediction (probability of tap) and updates the state of the input. So, the shape of this output is [1, 128].

## Threshold on Predictions

Predictions from the machine learning model are probabilistic, i.e., in the range [0.0, 1.0].

Our threshold for prediction is in the range `[minPredictionThreshold, maxPredictionThreshold]`, currently [0.5, 1.0]. This means that a prediction should be at least 0.5 to even be considered for further evaluation.

This threshold is modulated by a tap sensitivity setting; see below.

## Modeling Jerk

If the acceleration is the z-direction is in the range `[minJerkThreshold, maxJerkThreshold]`, we classify this as a jerk. Currently, this range is [30, 60]. `jerkThreshold` is initialized to `minJerkThreshold`. This is modulated based on a tap sensitivity setting; see below.

## Tap Sensitivity

Tap sensitivity is a number in the range [0, 100].

Below is a description of how the jerkThreshold and predictionThreshold are modulated based on the tap sensitivity and motor state:

### Is the SmartDrive Motor OFF?

#### Jerk Threshold

When the motor is OFF, jerkThreshold is computed as:

`jerkThreshold = maxJerkThreshold - (maxJerkThreshold - minJerkThreshold) * (sensitivity / 100.0)` 

For a tap sensitivity of 100%, the jerkThreshold becomes: `60 - (60 - 30) * 100/100 = 60 - 30 = 30`

For a tap sensitivity of 50%, the jerkThreshold becomes: `60 - (60 - 30) * 50 / 100 = 60 - 30 * 0.5 = 60 - 15 = 45`. As you can see, the jerkThreshold has increased from 30 to 45, i.e., the sensitivity to jerky motion has lowered. 

#### Prediction Threshold

When the motor is OFF, predictionThreshold is computed as:

`predictionThreshold = maxPredictionThreshold - (maxPredictionThreshold - minPredictionThreshold) * (sensitivity / 100)`

For a tap sensitivity of 20%, the predictionThreshold becomes: `1.0 - (1.0 - 0.5) * (20 / 100) = 0.9`. So, for a low sensitivity of 20%, we classify motion as a "tap" only if the model predicts a tap with a probability in the range [0.9, 1.0].

### Is the SmartDrive Motor ON?

#### Jerk Threshold

When the motor is ON, jerkThreshold is computed as:

`jerkThreshold = (maxJerkThreshold - (maxJerkThreshold - minJerkThreshold) * (sensitivity / 100.0)) - jerkThresholdOnOffDiff`. 

When the motor is ON, jerkThreshold is lowered by this parameter: `jerkThresholdOnOffDiff`, currently set to 10.0

So, for a tap sensitivity of 100%, the jerkThreshold is now 20 instead of 30 when the motor is ON. So, any acceleration in the z-direction in the range `[20, 60]` will likely be classified as a tap. 

#### Prediction Threshold

When the motor is ON, predictionThreshold is computed as:

`predictionThreshold = (maxPredictionThreshold - (maxPredictionThreshold - minPredictionThreshold) * (sensitivity / 100)) - predictionThresholdOnOffDiff`

When the motor is ON, predictionThreshold is lowered by this parameter: `predictionThresholdOnOffDiff`, currently set to 0.2.

So, for a tap sensitivity of 20%, the predictionThreshold is now 0.7 instead of 0.9 when the motor is ON. So, any prediction in the range [0.7, 1.0] will likely be classified as a tap.

## Tap Detection

Tap detection is a function that takes two arguments:

1. Current acceleration
2. Timestamp

and returns a single boolean - true if a tap is detected and false if not.

### Predict with Machine Learning

We first pass the input data to the tflite tap detector model and obtain a prediction. Then, we record this prediction in predictionHistory and update the model state.


### Return early

Now, we need to check based on the prediction if there was actually a tap:

We keep track of the last timestamp when a tap was detected - `lastTapTime` and check if it has been long enough to detect another tap. Currently, this "lockout" time is set to 100ms. This check is used to detect and prevent high-frequency tapping. 

Similarly, we have constraints on the size (/amount) of historical data we have seen so far. Currently, we record but do not detect the first 4 inputs (InputHistorySize) - this equates to the first 2 taps (PredictionHistorySize). 

### Analyze prediction history

Next, we find the minimum (minZ) and maximum (maxZ) acceleration in the z-direction that we have in recorded history. 

We calculate jerk as `jerk = maxZ - minZ` and store this in `jerkHistory`. We then find the maxJerk in our jerkHistory and check if it exceeds jerkThreshold. jerkThreshold is calculated based on tap sensitivity as noted above. This will tell us if the jerk is above our jerkThreshold, i.e., potentially a tap. 

Now, we check our predictionHistory to see if ALL predictions in our predictionHistory are above our predictionThreshold. Similar to jerkThreshold, predictionThreshold is set based on tap sensitivity as noted above.

### Inference rules

Our inference rule from prediction is this:

`predictTap = isJerkAboveThreshold and isPredictionAboveThreshold and tapGapObserved`

Read predictTap as "Is there a tap based on prediction?"

Here, if tapGap (boolean) is true, it indicates that we have a short period of time where the tap prediction was below the predictionThreshold (as seen in predictionHistory) before rising again above the threshold - Essentially, we're trying to catch the rising edge

Additional case:

If our maximum jerk (maxJerk) is greater than the maxJerkThreshold (currently 60) and the prediction is over the minimum prediction threshold (currently 0.5), this too "must have been a tap"

`mustTap = maxJerk > maxJerkThreshold and prediction > minPredictionThreshold`

Read mustTap as "Is there a tap based on jerk?"

Composition rule for tap detection: 

`Do we have a tap or not? = predictTap OR mustTap`

**Comment by xing-w** [comment]
> Our tap detector machine learning model takes 2 inputs:
> 
> Current input, as seen above
> Previous state (History) - An array of floats of size TapDetector.StateSize (currently set to 128) of predictions made by the tap detector so far.
> 
> So, the shape of this input is [3, 128].

The input should be [1,3] + [1,128], not [3,128].

**Comment by finger563** [comment]
Thanks @p-ranav @xing-w  and @guo-max - the documentation looks good and I think the code / implementation is miles ahead of where it was.

**Comment by finger563** [comment]
@p-ranav  can you approve the changes you requested :)
